sdk/ai/ai-inference-rest/src/trace.ts

```diff
@@ -0,0 +1,360 @@
+  messages.forEach((message: any) => {
+    if (message.role) {
+      let content: {
+        content?: string;
+        tool_calls?: Array<ChatCompletionsToolCallOutput>;
+        id?: string;
+      } = {};
+
+      const chatMsg = message as ChatRequestSystemMessage;
+      if (chatMsg.content) {
+        content.content = chatMsg.content;
+      }
+      if (!isContentRecordingEnabled()) {
+        content.content = "";
+      }
+
+      const assistantMsg = message as ChatRequestAssistantMessage;
+      if (assistantMsg.tool_calls) {
+        content.tool_calls = assistantMsg.tool_calls;
+        if (!isContentRecordingEnabled()) {
+          const toolCalls: Array<ChatCompletionsToolCallOutput> = JSON.parse(
+            JSON.stringify(content.tool_calls),
+          );
+          toolCalls.forEach((toolCall) => {
+            if (toolCall.function.arguments) {
+              toolCall.function.arguments = "";
+            }
+            toolCall.function.name = "";
+          });
+          content.tool_calls = toolCalls;
+        }
+      }
+
+      const toolMsg: ChatRequestToolMessage = message;
+      if (toolMsg.tool_call_id) {
+        content.id = toolMsg.tool_call_id;
+      }
+
+      span.addEvent(`gen_ai.${message.role}.message`, {
+        "gen_ai.system": INFERENCE_GEN_AI_SYSTEM_NAME,
+        "gen_ai.event.content": JSON.stringify(content),
+      });
+    }
+  });
+}
+
+/*
+* Add event to span.  Sample:
+{
+  name: 'gen_ai.choice',
+  attributes: {
+    'gen_ai.system': 'INFERENCE_GEN_AI_SYSTEM_NAME',
+    'gen_ai.event.content': '{"finish_reason":"tool_calls","index":0,"message":{"content":""}}'
+  },
+  time: [ 1725666881, 780608000 ],
+  droppedAttributesCount: 0
+}  
+*/
+function addResponseChatMessageEvent(span: TracingSpan, response: PathUncheckedResponse) {
+  response.body?.choices?.forEach((choice: ChatChoiceOutput) => {
+    let attributes;
+    let message: { content?: string; toolCalls?: Array<ChatCompletionsToolCallOutput> } = {};
+
+    if (choice.message.content) {
+      message.content = choice.message.content;
+    }
+    if (choice.message.tool_calls) {
+      message.toolCalls = choice.message.tool_calls;
+    }
+
+    if (!isContentRecordingEnabled()) {
+      message = JSON.parse(JSON.stringify(message));
+      message.content = "";
+      if (message.toolCalls) {
+        message.toolCalls.forEach((toolCall) => {
+          if (toolCall.function.arguments) {
+            toolCall.function.arguments = "";
+          }
+          toolCall.function.name = "";
+        });
+      }
+    }
+
+    const response = {
+      finish_reason: choice.finish_reason,
+      index: choice.index,
+      message,
+    };
+    attributes = {
+      "gen_ai.system": INFERENCE_GEN_AI_SYSTEM_NAME,
+      "gen_ai.event.content": JSON.stringify(response),
+    };
+
+    span.addEvent("gen_ai.choice", attributes);
+  });
+}
+
+export const tracingPolicy: PipelinePolicy = {
```

Comment:
Here's your custom policy - you now have access to:

1. The pipeline request
2. The pipeline response (if you `await next()` )
3. The tracing client 
4. Any package specific details

Change it however you'd like. The downside is that you're now working at a lower level so you only have the raw request body / headers / etc. You'll need to do some work to get it to work for your scenario but _everything_ should be available